src/useObservable.ts

```diff
@@ -33,24 +38,40 @@ export function useObservable<T>(observable: Observable<T>): T | undefined
 export function useObservable<T>(observable: Observable<T>, initialValue: T): T
 export function useObservable<T>(observable: Observable<T>, initialValue: () => T): T
 export function useObservable<T>(observable: Observable<T>, initialValue?: T | (() => T)) {
-  const [getSnapshot, subscribe] = useMemo(() => {
+  const [getSnapshot, subscribe] = useMemo<
+    [() => T, Parameters<typeof useSyncExternalStore>[0]]
+  >(() => {
     const record = getOrCreateStore(observable, getValue(initialValue))!
     return [
       function getSnapshot() {
+        // @TODO: perf opt opportunity: we could do `record.subscription.unsubscribe()` here to clear up some memory, as this subscription is only needed to provide a sync initialValue.
         return record.currentValue
       },
-      function subscribe(callback: (value: T) => void) {
-        const sub = record.observable.subscribe(next => callback(next))
+      function subscribe(callback: () => void) {
+        // @TODO: perf opt opportunity: we could do `record.subscription.unsubscribe()` here as we only need 1 subscription active to keep the observer alive
+        const sub = record.observable.subscribe(() => callback())
         return () => {
           sub.unsubscribe()
         }
       },
     ]
   }, [observable])
 
-  useIsomorphicEffect(() => {
+  const shouldRestoreSubscriptionRef = useRef(false)
+  useEffect(() => {
+    const store = getOrCreateStore(observable, getValue(initialValue))!
+    if (shouldRestoreSubscriptionRef.current) {
+      if (store.subscription.closed) {
```

Comment:
These two checks ensures that the hook can be called out of order, and enforces a `subscribe => unsubscribe => subscribe => unsubscribe` symmetry no matter what React might decide doing in the new v18 features.